gemd/entity/object/process_spec.py

```diff
@@ -56,31 +57,53 @@ class ProcessSpec(BaseObject, HasParameters, HasConditions, HasTemplate):
     def __init__(self, name, *, template=None,
                  parameters=None, conditions=None,
                  uids=None, tags=None, notes=None, file_links=None):
+        from gemd.entity.object.ingredient_spec import IngredientSpec
+        from gemd.entity.link_by_uid import LinkByUID
+
         BaseObject.__init__(self, name=name, uids=uids, tags=tags, notes=notes,
                             file_links=file_links)
         HasParameters.__init__(self, parameters=parameters)
         HasConditions.__init__(self, conditions=conditions)
-
-        self._ingredients = []
+        HasTemplate.__init__(self, template=template)
 
         # By default, a ProcessSpec is not linked to any MaterialSpec.
         # If a MaterialSpec is linked to this ProcessSpec,
         # then the field self._output_material will be automatically populated
         self._output_material = None
-
-        HasTemplate.__init__(self, template=template)
+        self._ingredients = validate_list(None, [IngredientSpec, LinkByUID])
 
     @property
     def ingredients(self):
         """Get the list of input ingredient specs."""
         return self._ingredients
 
-    def _unset_ingredient(self, ingred):
-        """Remove `ingred` from this process's list of ingredients."""
-        if ingred in self._ingredients:
-            self._ingredients.remove(ingred)
-
     @property
     def output_material(self):
         """Get the output material spec."""
         return self._output_material
+
+    # To avoid infinite recursion, fast return on revisit
+    global eq_seen
```

Comment:
Is this just a global because you can't change the signature of `__eq__`?  Why put the global in the class instead of outside of it?

Reply:
Yes; you can't change the sig of __eq__ and flow of control moves between several modules / is re-entrant.  I wish there was a way to get a tighter scope, but a closure wouldn't get ahead of it.  It's possible there's a better thread/exception safe approach, but I was also trying to maintain velocity.

Not opposed to moving the declaration to the top of file, but wrote it this way to be close to where it is used in the file.

Reply:
Better approach implemented, using the object itself as the state vector.